dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+                   make_string("The size of the sample of idx ", sample_idx, " is ", in_extents,
+                               ", which exceeds the limit of ", max_sample_extents, "."));
+    }
+  }
+
+  void SetupSampleDescs(int& max_total_workspace, const TensorListView<StorageGPU, Out, ndim>& out,
+                        const TensorListView<StorageGPU, const In, ndim>& in,
+                        const TensorListView<StorageGPU, const W, axes>& filters,
+                        const span<const ivec<axes>> anchors) {
+    const auto& out_shapes = out.shape;
+    const auto& in_shapes = in.shape;
+    const auto& filter_shapes = filters.shape;
+    int num_samples = in_shapes.num_samples();
+    ValidateNumericLimits(in_shapes, filter_shapes);
+    samples_desc_.clear();
+    samples_desc_.reserve(num_samples);
+    block_setups_.clear();
+    block_setups_.reserve(num_samples);
+    max_total_workspace = 0;
+    const int shared_mem_limit = GetSharedMemPerBlock();
+    for (int sample_idx = 0; sample_idx < num_samples; sample_idx++) {
+      int required_workspace;
+      BlockSetupT block_setup;
+      auto shape_desc = SetupSampleShapeDesc(
+          required_workspace, block_setup, sample_idx, out_shapes[sample_idx],
+          in_shapes[sample_idx], filter_shapes[sample_idx], anchors[sample_idx], shared_mem_limit);
+      max_total_workspace = std::max(max_total_workspace, required_workspace);
+      block_setups_.push_back(block_setup);
+      samples_desc_.push_back({out.tensor_data(sample_idx), in.tensor_data(sample_idx),
+                               filters.tensor_data(sample_idx), shape_desc});
+    }
+  }
+
+  template <typename InOutShape, typename FilterShape>
+  filter::ShapeDesc<axes> SetupSampleShapeDesc(int& required_workspace, BlockSetupT& block_setup,
+                                               int sample_idx, const InOutShape& out_shape,
+                                               const InOutShape& in_shape,
+                                               const FilterShape& filter_shape, ivec<axes> anchor,
+                                               int shared_mem_limit) {
+    int num_frames = has_sequence_dim ? in_shape[sequence_dim] : 1;
+    int num_channels = has_channel_dim ? in_shape[channels_dim] : 1;
+    const auto channels = cat(ivec<1>{num_channels}, ivec<axes - 1>{1});
+    auto in_extents = ShapeAsVec<int>(in_shape);
+    int width = in_extents.x;
+    auto filter_extents = shape2vec(filter_shape);
+    block_setup = PrepareBlockSetup(in_extents * channels);
+    int lanes_axis = GetLanesAxis(filter_extents);
+    auto log_block_extents = block_setup.block_dim();
+    log_block_extents[lanes_axis] *= StaticConfigT::lanes;
+    auto in_workspace_extents = (filter_extents - 1) * channels + log_block_extents;
+    auto in_workspace_num_elements = volume(static_cast<i64vec<axes>>(in_workspace_extents));
+    auto idx_workspace_size =
```

Comment:
These aren't long or complicated types - I'd appreciate fewer `auto` and more concrete types here.

Reply:
Done